### PR TITLE
WIP: Initial working of autcompletion in rcons

### PR DESCRIPTION
**Pull request by deepakchethan**
* The adding of newlines part when the cursor is at the bottom of screen is remaining 
Screen Shot 2019-05-12 at 4 38 58 AM

### CODE REVIEW

File: libr/cons/cons.c [hunk 1]
```diff
@@ -359,6 +359,35 @@ R_API bool r_cons_is_breaked() {
 	return I.context->breaked;
 }
 
+R_API int r_cons_get_cur_line () {
+	int curline = 0;
+	#if __WINDOWS__
```
Comment (radare) [inline libr/cons/cons.c]: cpp directives are not indented

File: libr/cons/cons.c [hunk 2]
```diff
@@ -359,6 +359,35 @@ R_API bool r_cons_is_breaked() {
...
+	POINT point;
+		if (GetCursorPos(&point)) {
+			curline = point.y;
+		}
+	#endif
+	#if __UNIX__
```
Comment (radare) [inline libr/cons/cons.c]: same here

File: libr/cons/cons.c [hunk 3]
```diff
@@ -359,6 +359,35 @@ R_API bool r_cons_is_breaked() {
...
+		char buf[8];
+		struct termios save,raw;
+		tcgetattr(0,&save);
+		cfmakeraw(&raw);
+		tcsetattr(0,TCSANOW,&raw);
+		if (isatty(fileno(stdin)))
```
Comment (radare) [inline libr/cons/cons.c]: missing space beforer ( and start { in the same lilne

File: libr/cons/cons.c [hunk 4]
```diff
@@ -359,6 +359,35 @@ R_API bool r_cons_is_breaked() {
...
+	#if __WINDOWS__
+	POINT point;
+		if (GetCursorPos(&point)) {
+			curline = point.y;
+		}
+	#endif
+	#if __UNIX__
+		char buf[8];
+		struct termios save,raw;
+		tcgetattr(0,&save);
```
Comment (radare) [inline libr/cons/cons.c]: missing spaces before ( and space after ,

File: libr/cons/cons.c [hunk 5]
```diff
@@ -359,6 +359,35 @@ R_API bool r_cons_is_breaked() {
 	return I.context->breaked;
 }
 
+R_API int r_cons_get_cur_line () {
+	int curline = 0;
+	#if __WINDOWS__
+	POINT point;
+		if (GetCursorPos(&point)) {
```
Comment (radare) [inline libr/cons/cons.c]: space before (

File: libr/cons/cons.c [hunk 6]
```diff
@@ -359,6 +359,35 @@ R_API bool r_cons_is_breaked() {
...
+	#endif
+	#if __UNIX__
+		char buf[8];
+		struct termios save,raw;
+		tcgetattr(0,&save);
+		cfmakeraw(&raw);
+		tcsetattr(0,TCSANOW,&raw);
+		if (isatty(fileno(stdin)))
+		{
+			write(1,R_CONS_GET_CURSOR_POSITION,sizeof(R_CONS_GET_CURSOR_POSITION));
```
Comment (radare) [inline libr/cons/cons.c]: spacings

File: libr/cons/dietline.c [hunk 7]
```diff
@@ -652,7 +691,9 @@ R_API void r_line_autocomplete() {
 		}
 	}
 
-	if (I.prompt_type != R_LINE_PROMPT_DEFAULT) {
+	if (I.prompt_type != R_LINE_PROMPT_DEFAULT ||
```
Comment (radare) [inline libr/cons/dietline.c]: in the same line

**Comment by radare** [comment]
Add a config option in scr. To allow the user specify which dietline autocompletion wants to use. This doesn’t needs to be the default

For the name of the var i dont have any preference or name, so im open to proposals

**Comment by XVilka** [comment]
@radare it is already here:
```c
 	SETCB ("cfg.autocomplete.widget", "true", &cb_showautocompletewidget, "Show widget dropdown for autocomplete"); 
```

**Comment by radare** [comment]
tahts a too long config name and also uses dot when there's no other var at this level. so we need a better name

File: libr/cons/cons.c (hunk 1 above)
Comment (deepakchethan) [reply libr/cons/cons.c]: Removed indentation

File: libr/cons/cons.c (hunk 5 above)
Comment (deepakchethan) [reply libr/cons/cons.c]: Added space :)

File: libr/cons/dietline.c (hunk 7 above)
Comment (deepakchethan) [reply libr/cons/dietline.c]: Added them to the same line

**Comment by radare** [comment]
rename the eval var to scr.prompt.popup

File: libr/cons/cons.c [hunk 8]
```diff
@@ -359,6 +359,34 @@ R_API bool r_cons_is_breaked() {
...
+		if (GetCursorPos(&point)) {
+			curline = point.y;
+		}
+#endif
+#if __UNIX__
+		char buf[8];
+		struct termios save,raw;
+		tcgetattr (0,&save);
+		cfmakeraw (&raw);
+		tcsetattr (0,TCSANOW,&raw);
```
Comment (radare) [inline libr/cons/cons.c]: spaces after commas

File: libr/cons/cons.c [hunk 9]
```diff
@@ -359,6 +359,34 @@ R_API bool r_cons_is_breaked() {
...
+		if (isatty (fileno (stdin))){
+			write(1, R_CONS_GET_CURSOR_POSITION, sizeof (R_CONS_GET_CURSOR_POSITION));
```
Comment (radare) [inline libr/cons/cons.c]: prefix the line with (void) to clarify the return value is ignored

File: libr/cons/cons.c [hunk 10]
```diff
@@ -359,6 +359,34 @@ R_API bool r_cons_is_breaked() {
...
+			read (0, buf, sizeof(buf));
```
Comment (radare) [inline libr/cons/cons.c]: missing space before sizeof(buf) and also check for return value

File: libr/cons/cons.c (hunk 9 above)
Comment (radare) [inline libr/cons/cons.c]: use strlen instead of sizeof

File: libr/include/r_cons.h [hunk 11]
```diff
@@ -857,6 +864,7 @@ typedef struct r_selection_widget_t {
 	int w, h;
 	int scroll;
 	bool complete_common;
+	int direction;
```
Comment (radare) [inline libr/include/r_cons.h]: if we dont plan to support more than 2 directions maybe worth to use bool and remove the "enum"

File: libr/cons/cons.c (hunk 8 above)
Comment (deepakchethan) [reply libr/cons/cons.c]: Done

File: libr/cons/cons.c (hunk 9 above)
Comment (deepakchethan) [reply libr/cons/cons.c]: Done

File: libr/include/r_cons.h (hunk 11 above)
Comment (deepakchethan) [reply libr/include/r_cons.h]: I just thought, it would increase the "readability" of the code. I will change it to bool :)

**Comment by radare** [comment]
usually readability is nice with enum because this way the compiler will catch missing switch ccases and such, but for this case i think its fine to use bool with a meaningful varname.

**Pull request merged**